pegen/pegen.h

```diff
@@ -49,11 +54,12 @@ void *keyword_token(Parser *p, const char *val);
 
 void *CONSTRUCTOR(Parser *p, ...);
 
-#define LINE(arg) ((expr_ty)(arg))->lineno
-#define COL(arg) ((expr_ty)(arg))->col_offset
-#define ENDLINE(arg) ((expr_ty)(arg))->end_lineno
-#define ENDCOL(arg) ((expr_ty)(arg))->end_col_offset
-#define EXTRA(head, tail) LINE(head), COL(head), ENDLINE(tail), ENDCOL(tail), p->arena
+#define EXTRA_EXPR(head, tail) EXTRA(head, expr_type, tail, expr_type)
+#define EXTRA(head, head_type_func, tail, tail_type_func) head_type_func(head, 1, 1), \
+                                                          head_type_func(head, 1, 0), \
+                                                          tail_type_func(tail, 0, 1), \
+                                                          tail_type_func(tail, 0, 0), \
+                                                          p->arena
```

Comment:
I have another idea. Maybe we can use C preprocessor "token pasting" (`token1##token2`) instead of the helper signature that takes `head, line` args. So we could do something like
```
...
head_type_func##headline(head), \
head_type_func##headcol(head), \
tail_type_func##tailline(tail), \
tail_type_func##tailcol(tail), \
...
```
And the definitions would be e.g.
```
int expr_type_headline(expr_ty a) { return a->line; }
int expr_type_headcol(expr_ty a) { return a->col_offset; }
```
etc.

Or actually they could even be macros:
```
#define expr_type_headline(a) (a->line)
```
etc.

Reply:
@gvanrossum That indeed looks like a better idea! I'll implement it ASAP and the start testing to find out what goes wrong with the line/col info.

Is there maybe a way to globally disable memoization? In the rules I want to debug, it is not really needed and it makes debugging a bit more difficult for me, since I haven't still understood it 100%.

Reply:
You can disable memoization in the generator. At the top of visit_Rule() there's a line `memoize = not node.left_recursive`. Just set that to False and you should be good. However for left-recursive rules you'll still get use of the memo, the left-recursion algorithm depends on it. (But it doesn't use that flag, so just setting `memoize = False` should work.)